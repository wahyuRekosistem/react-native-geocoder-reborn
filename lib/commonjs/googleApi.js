"use strict";var _interopRequireDefault=require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports,"__esModule",{value:true});exports.default=void 0;var _regenerator=_interopRequireDefault(require("@babel/runtime/regenerator"));var GOOGLE_URL='https://maps.google.com/maps/api/geocode/json';function geocodeRequest(url,headers){var res,json;return _regenerator.default.async(function geocodeRequest$(_context){while(1){switch(_context.prev=_context.next){case 0:_context.next=2;return _regenerator.default.awrap(fetch(url,{headers:headers}));case 2:res=_context.sent;_context.next=5;return _regenerator.default.awrap(res.json());case 5:json=_context.sent;if(!(!json.results||json.status!=='OK')){_context.next=8;break;}throw new Error("Google Maps Error: ["+json.status+"] "+json.error_message);case 8:return _context.abrupt("return",json.results.map(format));case 9:case"end":return _context.stop();}}},null,null,null,Promise);}function format(raw){var geocodeObj={position:{},formattedAddress:raw.formatted_address||'',feature:null,streetNumber:null,streetName:null,postalCode:null,locality:null,country:null,countryCode:null,adminArea:null,subAdminArea:null,subLocality:null};if(raw.geometry&&raw.geometry.location){geocodeObj.position={lat:raw.geometry.location.lat,lng:raw.geometry.location.lng};}raw.address_components.forEach(function(component){if(component.types.indexOf('route')!==-1){geocodeObj.streetName=component.long_name;}else if(component.types.indexOf('street_number')!==-1){geocodeObj.streetNumber=component.long_name;}else if(component.types.indexOf('country')!==-1){geocodeObj.country=component.long_name;geocodeObj.countryCode=component.short_name;}else if(component.types.indexOf('locality')!==-1){geocodeObj.locality=component.long_name;}else if(component.types.indexOf('postal_code')!==-1){geocodeObj.postalCode=component.long_name;}else if(component.types.indexOf('administrative_area_level_1')!==-1){geocodeObj.adminArea=component.long_name;}else if(component.types.indexOf('administrative_area_level_2')!==-1){geocodeObj.subAdminArea=component.long_name;}else if(component.types.indexOf('sublocality')!==-1||component.types.indexOf('sublocality_level_1')!==-1){geocodeObj.subLocality=component.long_name;}else if(component.types.indexOf('point_of_interest')!==-1||component.types.indexOf('colloquial_area')!==-1){geocodeObj.feature=component.long_name;}});return geocodeObj;}var _default={geocodePosition:function geocodePosition(apiKey,position,language,headers){return geocodeRequest(GOOGLE_URL+"?key="+apiKey+"&latlng="+position.lat+","+position.lng+"&language="+language,headers);},geocodeAddress:function geocodeAddress(apiKey,address,language,headers){return geocodeRequest(GOOGLE_URL+"?key="+apiKey+"&address="+encodeURI(address)+"&language="+language,headers);},geocodeAddressWithBounds:function geocodeAddressWithBounds(apiKey,address,bounds,language,headers){var sw=bounds.sw,ne=bounds.ne;return geocodeRequest(GOOGLE_URL+"?key="+apiKey+"&address="+encodeURI(address)+("&bounds="+sw.lat+","+sw.lng+"|"+ne.lat+","+ne.lng+"&language="+language),headers);}};exports.default=_default;
//# sourceMappingURL=googleApi.js.map