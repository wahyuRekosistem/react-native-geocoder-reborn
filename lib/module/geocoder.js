var _interopRequireDefault=require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports,"__esModule",{value:true});exports.default=void 0;var _regenerator=_interopRequireDefault(require("@babel/runtime/regenerator"));var _reactNative=require("react-native");var _native=_interopRequireDefault(require("./native"));var _googleApi=_interopRequireDefault(require("./googleApi"));var defaultOptions={locale:'en',fallbackToGoogle:false,forceGoogleOnIos:false,maxResults:2};function getApiKey(apiKey){if(!apiKey){throw new Error('Invalid API Key: `apiKey` is required for using Google Maps API.');}return apiKey;}function geocodePositionGoogle(position){var options,apiKey,locale,headers,_args=arguments;return _regenerator.default.async(function geocodePositionGoogle$(_context){while(1){switch(_context.prev=_context.next){case 0:options=_args.length>1&&_args[1]!==undefined?_args[1]:{};apiKey=getApiKey(options.apiKey);locale=options.locale||defaultOptions.locale;headers=options.requestHeaders;return _context.abrupt("return",_googleApi.default.geocodePosition(apiKey,position,locale,headers));case 5:case"end":return _context.stop();}}},null,null,null,Promise);}function geocodeAddressGoogle(address){var options,apiKey,locale,headers,_args2=arguments;return _regenerator.default.async(function geocodeAddressGoogle$(_context2){while(1){switch(_context2.prev=_context2.next){case 0:options=_args2.length>1&&_args2[1]!==undefined?_args2[1]:{};apiKey=getApiKey(options.apiKey);locale=options.locale||defaultOptions.locale;headers=options.requestHeaders;if(!options.bounds){_context2.next=6;break;}return _context2.abrupt("return",_googleApi.default.geocodeAddressWithBounds(apiKey,address,options.bounds,locale,headers));case 6:return _context2.abrupt("return",_googleApi.default.geocodeAddress(apiKey,address,locale,headers));case 7:case"end":return _context2.stop();}}},null,null,null,Promise);}function geocodePosition(position){var options,_args3=arguments;return _regenerator.default.async(function geocodePosition$(_context3){while(1){switch(_context3.prev=_context3.next){case 0:options=_args3.length>1&&_args3[1]!==undefined?_args3[1]:{};if(!(!position||position.lat==null||position.lng==null)){_context3.next=3;break;}throw new Error('Invalid Position: `{lat, lng}` is required');case 3:if(!(options.forceGoogleOnIos&&_reactNative.Platform.OS==='ios')){_context3.next=5;break;}return _context3.abrupt("return",geocodePositionGoogle(position,options));case 5:if(!(_native.default==null)){_context3.next=9;break;}if(!options.fallbackToGoogle){_context3.next=8;break;}return _context3.abrupt("return",geocodePositionGoogle(position,options));case 8:throw new Error('Missing Native Module: Please check the module linking, '+'or set `fallbackToGoogle` in the init options.');case 9:_context3.prev=9;if(!(_reactNative.Platform.OS==='android')){_context3.next=16;break;}_context3.next=13;return _regenerator.default.awrap(_native.default.geocodePositionAndroid(position,options));case 13:return _context3.abrupt("return",_context3.sent);case 16:_context3.next=18;return _regenerator.default.awrap(_native.default.geocodePosition(position,options.locale||defaultOptions.locale,options.maxResults||defaultOptions.maxResults));case 18:return _context3.abrupt("return",_context3.sent);case 19:_context3.next=26;break;case 21:_context3.prev=21;_context3.t0=_context3["catch"](9);if(!options.fallbackToGoogle){_context3.next=25;break;}return _context3.abrupt("return",geocodePositionGoogle(position,options));case 25:throw new Error('Native Error: '+(_context3.t0===null||_context3.t0===void 0?void 0:_context3.t0.message)||'Unknown Execption.');case 26:case"end":return _context3.stop();}}},null,null,[[9,21]],Promise);}function geocodeAddress(address){var options,_options$regionIos,center,radius,_args4=arguments;return _regenerator.default.async(function geocodeAddress$(_context4){while(1){switch(_context4.prev=_context4.next){case 0:options=_args4.length>1&&_args4[1]!==undefined?_args4[1]:{};if(address){_context4.next=3;break;}throw new Error('Invalid Address: `string` is required');case 3:if(!(options.forceGoogleOnIos&&_reactNative.Platform.OS==='ios')){_context4.next=5;break;}return _context4.abrupt("return",geocodeAddressGoogle(address,options));case 5:if(!(_native.default==null)){_context4.next=9;break;}if(!options.fallbackToGoogle){_context4.next=8;break;}return _context4.abrupt("return",geocodeAddressGoogle(address,options));case 8:throw new Error('Missing Native Module: Please check the module linking, '+'or set `fallbackToGoogle` in the init options.');case 9:_context4.prev=9;if(!(_reactNative.Platform.OS==='ios'&&options.regionIos)){_context4.next=17;break;}_options$regionIos=options.regionIos,center=_options$regionIos.center,radius=_options$regionIos.radius;_context4.next=14;return _regenerator.default.awrap(_native.default.geocodeAddressInRegion(address,center.lat,center.lng,radius,options.locale||defaultOptions.locale,options.maxResults||defaultOptions.maxResults));case 14:return _context4.abrupt("return",_context4.sent);case 17:if(!(_reactNative.Platform.OS==='android')){_context4.next=23;break;}_context4.next=20;return _regenerator.default.awrap(_native.default.geocodeAddressAndroid(address,options));case 20:return _context4.abrupt("return",_context4.sent);case 23:_context4.next=25;return _regenerator.default.awrap(_native.default.geocodeAddress(address,options.locale||defaultOptions.locale,options.maxResults||defaultOptions.maxResults));case 25:return _context4.abrupt("return",_context4.sent);case 26:_context4.next=33;break;case 28:_context4.prev=28;_context4.t0=_context4["catch"](9);if(!options.fallbackToGoogle){_context4.next=32;break;}return _context4.abrupt("return",geocodeAddressGoogle(address,options));case 32:throw new Error('Native Error: '+(_context4.t0===null||_context4.t0===void 0?void 0:_context4.t0.message)||'Unknown Execption.');case 33:case"end":return _context4.stop();}}},null,null,[[9,28]],Promise);}var _default={geocodePosition:geocodePosition,geocodePositionGoogle:geocodePositionGoogle,geocodeAddress:geocodeAddress,geocodeAddressGoogle:geocodeAddressGoogle};exports.default=_default;
//# sourceMappingURL=geocoder.js.map