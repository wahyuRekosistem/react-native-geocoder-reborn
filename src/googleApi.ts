import { Position, Bounds, GeocodingObject, RequestHeaders } from './types';

const GOOGLE_URL = 'https://maps.google.com/maps/api/geocode/json';

async function geocodeRequest(url: string, headers?: RequestHeaders) {
  let res = await fetch(url, { headers });
  let json = await res.json();

  if (!json.results || json.status !== 'OK') {
    throw new Error(
      `Google Maps Error: [${json.status}] ${json.error_message}`
    );
  }
  return json.results.map(format) as GeocodingObject[];
}

function format(raw: any) {
  const geocodeObj = {
    position: {},
    formattedAddress: raw.formatted_address || '',
    feature: null,
    streetNumber: null,
    streetName: null,
    postalCode: null,
    locality: null,
    country: null,
    countryCode: null,
    adminArea: null,
    subAdminArea: null,
    subLocality: null,
  };

  if (raw.geometry && raw.geometry.location) {
    geocodeObj.position = {
      lat: raw.geometry.location.lat,
      lng: raw.geometry.location.lng,
    };
  }

  raw.address_components.forEach((component: any) => {
    if (component.types.indexOf('route') !== -1) {
      geocodeObj.streetName = component.long_name;
    } else if (component.types.indexOf('street_number') !== -1) {
      geocodeObj.streetNumber = component.long_name;
    } else if (component.types.indexOf('country') !== -1) {
      geocodeObj.country = component.long_name;
      geocodeObj.countryCode = component.short_name;
    } else if (component.types.indexOf('locality') !== -1) {
      geocodeObj.locality = component.long_name;
    } else if (component.types.indexOf('postal_code') !== -1) {
      geocodeObj.postalCode = component.long_name;
    } else if (component.types.indexOf('administrative_area_level_1') !== -1) {
      geocodeObj.adminArea = component.long_name;
    } else if (component.types.indexOf('administrative_area_level_2') !== -1) {
      geocodeObj.subAdminArea = component.long_name;
    } else if (
      component.types.indexOf('sublocality') !== -1 ||
      component.types.indexOf('sublocality_level_1') !== -1
    ) {
      geocodeObj.subLocality = component.long_name;
    } else if (
      component.types.indexOf('point_of_interest') !== -1 ||
      component.types.indexOf('colloquial_area') !== -1
    ) {
      geocodeObj.feature = component.long_name;
    }
  });

  return geocodeObj;
}

export default {
  geocodePosition(
    apiKey: string,
    position: Position,
    language: string,
    headers?: RequestHeaders
  ) {
    return geocodeRequest(
      `${GOOGLE_URL}?key=${apiKey}&latlng=${position.lat},${position.lng}&language=${language}`,
      headers
    );
  },

  geocodeAddress(
    apiKey: string,
    address: string,
    language: string,
    headers?: RequestHeaders
  ) {
    return geocodeRequest(
      `${GOOGLE_URL}?key=${apiKey}&address=${encodeURI(
        address
      )}&language=${language}`,
      headers
    );
  },

  geocodeAddressWithBounds(
    apiKey: string,
    address: string,
    bounds: Bounds,
    language: string,
    headers?: RequestHeaders
  ) {
    let { sw, ne } = bounds;
    return geocodeRequest(
      `${GOOGLE_URL}?key=${apiKey}&address=${encodeURI(address)}` +
        `&bounds=${sw.lat},${sw.lng}|${ne.lat},${ne.lng}&language=${language}`,
      headers
    );
  },
};
