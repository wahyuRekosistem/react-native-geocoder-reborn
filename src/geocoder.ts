import { Platform } from 'react-native';
import nativeImpl from './native';
import googleApi from './googleApi';
import { Position, GeocoderOptions, GeocodingObject } from './types';

let defaultOptions = {
  locale: 'en',
  fallbackToGoogle: false,
  forceGoogleOnIos: false,
  maxResults: 2,
};

function getApiKey(apiKey?: string) {
  if (!apiKey) {
    throw new Error(
      'Invalid API Key: `apiKey` is required for using Google Maps API.'
    );
  }
  return apiKey;
}

async function geocodePositionGoogle(
  position: Position,
  options: GeocoderOptions = {}
) {
  let apiKey = getApiKey(options.apiKey);
  let locale = options.locale || defaultOptions.locale;
  let headers = options.requestHeaders;
  return googleApi.geocodePosition(apiKey, position, locale, headers);
}

async function geocodeAddressGoogle(
  address: string,
  options: GeocoderOptions = {}
) {
  let apiKey = getApiKey(options.apiKey);
  let locale = options.locale || defaultOptions.locale;
  let headers = options.requestHeaders;
  if (options.bounds) {
    // Use rectangle bounds for Google Maps api
    return googleApi.geocodeAddressWithBounds(
      apiKey,
      address,
      options.bounds,
      locale,
      headers
    );
  }
  return googleApi.geocodeAddress(apiKey, address, locale, headers);
}

async function geocodePosition(
  position: Position,
  options: GeocoderOptions = {}
): Promise<GeocodingObject[]> {
  if (!position || position.lat == null || position.lng == null) {
    throw new Error('Invalid Position: `{lat, lng}` is required');
  }

  if (options.forceGoogleOnIos && Platform.OS === 'ios') {
    return geocodePositionGoogle(position, options);
  }

  if (nativeImpl == null) {
    if (options.fallbackToGoogle) {
      return geocodePositionGoogle(position, options);
    }
    throw new Error(
      'Missing Native Module: Please check the module linking, ' +
        'or set `fallbackToGoogle` in the init options.'
    );
  }

  try {
    if (Platform.OS === 'android') {
      return await nativeImpl.geocodePositionAndroid(position, options);
    } else {
      return await nativeImpl.geocodePosition(
        position,
        options.locale || defaultOptions.locale,
        options.maxResults || defaultOptions.maxResults
      );
    }
  } catch (err) {
    if (options.fallbackToGoogle) {
      return geocodePositionGoogle(position, options);
    }
    throw new Error('Native Error: ' + err?.message || 'Unknown Execption.');
  }
}

async function geocodeAddress(
  address: string,
  options: GeocoderOptions = {}
): Promise<GeocodingObject[]> {
  if (!address) {
    throw new Error('Invalid Address: `string` is required');
  }

  if (options.forceGoogleOnIos && Platform.OS === 'ios') {
    return geocodeAddressGoogle(address, options);
  }

  if (nativeImpl == null) {
    if (options.fallbackToGoogle) {
      return geocodeAddressGoogle(address, options);
    }
    throw new Error(
      'Missing Native Module: Please check the module linking, ' +
        'or set `fallbackToGoogle` in the init options.'
    );
  }

  try {
    if (Platform.OS === 'ios' && options.regionIos) {
      // Use round region query for iOS
      let { center, radius } = options.regionIos;
      return await nativeImpl.geocodeAddressInRegion(
        address,
        center.lat,
        center.lng,
        radius,
        options.locale || defaultOptions.locale,
        options.maxResults || defaultOptions.maxResults
      );
    } else if (Platform.OS === 'android') {
      return await nativeImpl.geocodeAddressAndroid(address, options);
    } else {
      return await nativeImpl.geocodeAddress(
        address,
        options.locale || defaultOptions.locale,
        options.maxResults || defaultOptions.maxResults
      );
    }
  } catch (err) {
    if (options.fallbackToGoogle) {
      return geocodeAddressGoogle(address, options);
    }
    throw new Error('Native Error: ' + err?.message || 'Unknown Execption.');
  }
}

export default {
  geocodePosition,
  geocodePositionGoogle,
  geocodeAddress,
  geocodeAddressGoogle,
};
