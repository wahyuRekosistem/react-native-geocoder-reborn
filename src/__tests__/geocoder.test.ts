import geocoder from '../geocoder';
import nativeImpl from '../native';
import googleApi from '../googleApi';

jest.mock('../native', () => ({
  init: jest.fn(),
  geocodePosition: jest.fn(),
  geocodeAddress: jest.fn(),
  geocodeAddressWithBounds: jest.fn(),
}));

jest.mock('../googleApi', () => ({
  geocodePosition: jest.fn(),
  geocodeAddress: jest.fn(),
  geocodeAddressWithBounds: jest.fn(),
}));

describe('geocode position', () => {
  it('should throw if position is invalid', async () => {
    // @ts-ignore: Expect to throw
    await expect(geocoder.geocodePosition({})).rejects.toThrowError();
  });

  it('should call native function with position', async () => {
    let position = { lat: -1.2, lng: 3.4 };
    await geocoder.geocodePosition(position);
    expect(nativeImpl.geocodePosition).toMatchSnapshot();
  });

  it('should throw if native function rejects and fallbackToGoogle is not set', async () => {
    nativeImpl.geocodePosition = () => Promise.reject(new Error('NO IMPL'));
    let position = { lat: -1.2, lng: 3.4 };
    await expect(geocoder.geocodePosition(position)).rejects.toThrowError();
  });

  it('should call googleApi function with position if fallbackToGoogle is set and native fails', async () => {
    nativeImpl.geocodePosition = () => Promise.reject(new Error('NO IMPL'));
    let position = { lat: -1.2, lng: 3.4 };
    let options = {
      fallbackToGoogle: true,
      apiKey: 'TEST',
      locale: 'de',
    };
    await geocoder.geocodePosition(position, options);
    expect(googleApi.geocodePosition).toMatchSnapshot();
  });
});

describe('geocode address', () => {
  it('should throw if address is invalid', async () => {
    expect(geocoder.geocodeAddress('')).rejects.toThrowError();
  });

  it('should throw if native function rejects and fallbackToGoogle is not set', async () => {
    nativeImpl.geocodeAddress = () => Promise.reject(new Error('NO IMPL'));
    let address = 'London';
    let options = { fallbackToGoogle: false };
    await expect(
      geocoder.geocodeAddress(address, options)
    ).rejects.toThrowError();
  });

  it('should call googleApi function with address if fallbackToGoogle is set and native fails', async () => {
    nativeImpl.geocodeAddress = () => Promise.reject(new Error('NO IMPL'));
    let address = 'London';
    let options = {
      fallbackToGoogle: true,
      apiKey: 'TEST',
      locale: 'en',
    };
    await geocoder.geocodeAddress(address, options);
    expect(googleApi.geocodeAddress).toMatchSnapshot();
  });

  it('should call googleApi function with address and bounds', async () => {
    nativeImpl.geocodeAddress = () => Promise.reject(new Error('NO IMPL'));
    let address = 'London';
    let bounds = { sw: { lat: 40, lng: -20 }, ne: { lat: 60, lng: 10 } };
    let options = {
      fallbackToGoogle: true,
      apiKey: 'TEST',
      locale: 'en',
      bounds,
    };
    await geocoder.geocodeAddress(address, options);
    expect(googleApi.geocodeAddressWithBounds).toMatchSnapshot();
  });
});
