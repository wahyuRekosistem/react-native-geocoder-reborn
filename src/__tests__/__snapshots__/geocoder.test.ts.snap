// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geocode address should call googleApi function with address and bounds 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "TEST",
      "London",
      Object {
        "ne": Object {
          "lat": 60,
          "lng": 10,
        },
        "sw": Object {
          "lat": 40,
          "lng": -20,
        },
      },
      "en",
      undefined,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`geocode address should call googleApi function with address if fallbackToGoogle is set and native fails 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "TEST",
      "London",
      "en",
      undefined,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`geocode position should call googleApi function with position if fallbackToGoogle is set and native fails 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "TEST",
      Object {
        "lat": -1.2,
        "lng": 3.4,
      },
      "de",
      undefined,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`geocode position should call native function with position 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "lat": -1.2,
        "lng": 3.4,
      },
      "en",
      2,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
